src/TDB2.cpp

```diff
@@ -42,967 +42,161 @@
-TDB2::TDB2 ()
-: _location ("")
-, _id (1)
-{
-  // Mark the pending file as the only one that has ID numbers.
-  pending.has_ids ();
-
-  // Indicate that dependencies should be automatically scanned on startup,
-  // setting Task::is_blocked and Task::is_blocking accordingly.
-  pending.auto_dep_scan ();
-}
-
-////////////////////////////////////////////////////////////////////////////////
-// Once a location is known, the files can be set up.  Note that they are not
-// read.
-void TDB2::set_location (const std::string& location)
-{
-  _location = location;
-
-  pending.target   (location + "/pending.data");
-  completed.target (location + "/completed.data");
-  undo.target      (location + "/undo.data");
-  backlog.target   (location + "/backlog.data");
-}
-
-////////////////////////////////////////////////////////////////////////////////
-// Add the new task to the appropriate file.
-void TDB2::add (Task& task, bool add_to_backlog /* = true */)
-{
-  // Ensure the task is consistent, and provide defaults if necessary.
-  // bool argument to validate() is "applyDefault". Pass add_to_backlog through
-  // in order to not apply defaults to synchronized tasks.
-  task.validate (add_to_backlog);
-  std::string uuid = task.get ("uuid");
 
-  // If the tasks are loaded, then verify that this uuid is not already in
-  // the file.
-  if (!verifyUniqueUUID (uuid))
-    throw format ("Cannot add task because the uuid '{1}' is not unique.", uuid);
+  auto ws = replica.working_set ();
 
-  // Only locally-added tasks trigger hooks.  This means that tasks introduced
-  // via 'sync' do not trigger hooks.
-  if (add_to_backlog)
-    Context::getContext ().hooks.onAdd (task);
+  // get the ID that was assigned to this task
+  auto id = ws.by_uuid (uuid);
 
-  update (task, add_to_backlog, true);
-}
-
-////////////////////////////////////////////////////////////////////////////////
-void TDB2::modify (Task& task, bool add_to_backlog /* = true */)
-{
-  // Ensure the task is consistent.
-  task.validate (false);
-  std::string uuid = task.get ("uuid");
+  // update the cached working set with the new information
+  _working_set = std::make_optional (std::move (ws));
 
-  // Get the unmodified task as reference, so the hook can compare.
-  if (add_to_backlog)
-  {
-    Task original;
-    get (uuid, original);
-    Context::getContext ().hooks.onModify (original, task);
+  if (id.has_value ()) {
+    task.id = id.value();
   }
 
-  update (task, add_to_backlog);
+  // run hooks for this new task
+  Context::getContext ().hooks.onAdd (task);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-void TDB2::update (
-  Task& task,
-  const bool add_to_backlog,
-  const bool addition /* = false */)
+// Modify the task in storage to match the given task.
+//
+// This exhibits a bit of a race condition: if the stored task has changed since
+// it was loaded, this will revert those changes.  In practice, this is not an
+// issue.
```

Comment:
See my comment on "How could this task not exist".

Reply:
Thanks. Maybe you could incorporate the more concrete application in that comment into this code comment? Something like:

> This exhibits a bit of a race condition: if the stored task has changed since it was loaded, for example by two simultaneous `task` invocations, this will revert those changes. In practice, this is not an issue because we don't delete tasks until they expire.

Reply:
I'll add to the comment. Note that there are two different race conditions here, although they are related.
 * Multiple changes are made simultaneously.
 * A task being deleted simultaneously with a modification.

In the first case, whichever change commits last, wins. In the second case, the deletion wins.

Reply:
Well, the deletion _could_ win, but instead we throw an error.